### ISSUE TITLE
`TextInput` component not handling events during the initial life cycle

### ISSUE THREAD

**Issue by ylkhayat** [marmelab/react-admin#9670]
## Some bit of context

We use `react-admin` version `4.16.9` in a `pnpm` monorepo workspace. And we have the automatic renovate pull requests to update our repository packages.
Our end to end tests were successful before updating to `react-hook-form` from `7.44.3` to `7.51.0` then it started to fail consistently. The failing test was related to `react-admin`'s logging in page. 

#### Test is as follows (in this order):
- Cypress finds the `username` and `password` fields.
- Cypress starts typing in the corresponding credentials.

Assume the credentials are:
- `username`: `a-user`
- `password`: `what-a-password`

**What you were expecting:**
It was expected to fill in the `username` and `password` fields normally. But we noticed that the test was failing because the typed in username was `-user` (notice the missing `a`) but the password field was normal `what-a-password`.

**What happened instead:**
It was expected of course that the `username` and `password` to have `a-user` and `what-a-password` respectively.

**Steps to reproduce:**

You can check the reproduction code here [react-admin-delayed-sync](https://codesandbox.io/p/devbox/react-admin-delayed-sync-56pg7h).

If you change `react-hook-form` to `7.44.3` then you will see the sync works as expected

#### Version `7.44.3`

https://github.com/marmelab/react-admin/assets/29442110/9d568ab7-78d8-4035-9ae0-30105327c714

#### Version `7.50.1`

https://github.com/marmelab/react-admin/assets/29442110/034ad9a0-8300-4d86-b81e-ab16013b9746

Thus, we can notice that the issue is with the `TextInput` component not from the registered input component with `react-hook-form`.

**Related code:**

You can check the sandbox that I have created for this [react-admin-delayed-sync](https://codesandbox.io/p/devbox/react-admin-delayed-sync-56pg7h).

**Environment**

* React-admin version: ^4.16.10
* React version: 18.2.0
* React Hook Form version: 7.50.1
* Browser: Chrome

**Comment by adguernier** [comment]
Thanks for this report and the sandbox :pray: 

However, tweaking the reproductible [react-admin-sanbox](https://codesandbox.io/p/devbox/react-admin-delayed-sync-56pg7h) by moving the `document.querySelector('input[id="username-ra"]');` before the `useEffect` solves the issue you describe.
```js
const raInput = document.querySelector('input[id="username-ra"]'); // it works
  useEffect(() => {
      ...
```

Knowing that the React way to get an element is to use `useRef`, I wonder if using `document.querySelector` instead of a `ref` could cause this issue (although for now , `<TextInput>` does not propagate to `ref`). :thinking: 

All things considered, for me this use case is unreliable to properly investigate and qualify this issue.
May you consider providing a project repository using the Cypress tests which lead to the issue you describe? Thanks

**Comment by ylkhayat** [comment]
@adguernier you're absolutely correct, but I am not sure how did you make it work with a `ref`, that's definitely the first thing I have tried, so I had to hack it in somehow. I even tried with `inputRef` and it didn't work. Check the `console.log()` in the upcoming screenshot.

<img width="774" alt="image" src="https://github.com/marmelab/react-admin/assets/29442110/048bd8f6-31dd-4bdb-bcb3-4f502ca2a3c3">

And also regarding the moving this line one level up, it defies the purpose of the test, because the issue here is about handling the events during the very initial life cycles. If you simply move it up one level then you will start typing exactly one event later, this case is when the document becomes available but the `useEffect` didn't yet run. But by putting in the `querySelector` in the `useEffect` it's a dynamic fetch (kinda). 

Please have a look at the screen shots attached below.
<table>
<tr>
<th>
<h4> Inside useEffect </h4>
</th>
<th>
<h4> Outside useEffect </h4>
</th>
</tr>
<tr>
<td>

<img width="1761" alt="image" src="https://github.com/marmelab/react-admin/assets/29442110/1cb2362f-796f-4bb3-bae4-5adb047a5972">

</td>
<td>

<img width="1761" alt="image" src="https://github.com/marmelab/react-admin/assets/29442110/f06b4260-aed4-4c25-96e0-bc22247b2fea">

</td>
</tr>
</table>


Thanks for looking into this! And happy to help out!

**Comment by ylkhayat** [comment]
@fzaninotto But this (`LoginForm`) is a black box to us, we don't have access to the `LoginForm` provided from `react-admin` and we'd like to keep using the built in form, without overriding the `loginPage` prop passed to the `AdminUI` wrapper, we just use it as it is. 

Will this be handled from react-admin's side to simplify the `LoginForm` usage as you mentioned in an upcoming version?
